src/movies.js

```diff
@@ -1,16 +1,107 @@
 // Iteration 1: All directors? - Get the array of all directors.
+function getAllDirectors(movies) {
+    const movieDirector = movies.map(movie => movie.director)
+    return movieDirector
+}
 // _Bonus_: It seems some of the directors had directed multiple movies so they will pop up multiple times in the array of directors. How could you "clean" a bit this array and make it unified (without duplicates)?
 
 // Iteration 2: Steven Spielberg. The best? - How many drama movies did STEVEN SPIELBERG direct?
+function howManyMovies(movies) {
+    const moviesSpielberg = movies.filter(elm => elm.director === "Steven Spielberg" && elm.genre.includes("Drama"))
+    return moviesSpielberg.length
+}
 
 // Iteration 3: All rates average - Get the average of all rates with 2 decimals
+function ratesAverage(movies) {
+    if (movies.length === 0) {
+        return 0
+    } else {
+        const movieRateTotal = movies.reduce((acc, elm) => acc + elm.rate, 0)
+        const movieWithoutRate = movies.filter(elm => elm.rate === null || undefined || '') // I tried to take off the movies without rate but it doesn't seem working.
+        const movieRateAvg = (movieRateTotal-movieWithoutRate) / (movies.length-movieWithoutRate.length)
+        const roundedmovieRateAvg = Number(movieRateAvg.toFixed(2))
+        return roundedmovieRateAvg
+    }
+}
 
 // Iteration 4: Drama movies - Get the average of Drama Movies
+function dramaMoviesRate(movies) {
+    const dramamovie = movies.filter(elm => elm.genre.includes("Drama"))
+    if (dramamovie.length === 0) {
+        return 0
+    } else {
+        const totalDramaMovieRate = dramamovie.reduce((acc, elm) => acc + elm.rate, 0)
+        const avgDramaMovieRate = totalDramaMovieRate/dramamovie.length
+        const roundedDramaMovieRate = Number(avgDramaMovieRate.toFixed(2))
+        return roundedDramaMovieRate
+    }
+}
 
 // Iteration 5: Ordering by year - Order by year, ascending (in growing order)
+function orderByYear(movies) {
+    const moviescopy = JSON.parse(JSON.stringify(movies))
+    moviescopy.sort(function(a, b) {
+        if (a.year === b.year) { // I tried that for ordering the titles alphabetically if the years were the same but it doesn't seem working
+            moviescopy.sort((a, b) => {
+            let fa = a.title,
+            fb = b.title
+
+            if (fa < fb) {
+            return -1
+            }
+            if (fa > fb) {
+            return 1
+            }
+            return 0
+        }) 
+        } else {
+            return a.year - b.year
+        }
+    })
+    return moviescopy
+}        
```

Comment:
Has estado muuuuy cerca. El problema es que has intentado ordenar movies mientras estabas ordenando movies... lo cual es imposible.
```suggestion
    moviescopy.sort(function(a, b) {
        if (a.year === b.year) { // I tried that for ordering the titles alphabetically if the years were the same but it doesn't seem working
    
            let fa = a.title,
            fb = b.title

            if (fa < fb) {
            return -1
            } else if (fa > fb) {
            return 1
            } else {
            return 0
            }
        } else {
            return a.year - b.year
        }
    })
    return moviescopy
}        
```

Eso ya funcionaría. 

Creo que más o menos entiendes como va pero vamos a repasarlo.

Sort puede recibir una función callback que se comporta de la siguiente forma. Si retorna...

Un número positivo (1), coloca "b" antes que "a"
Un número negativo (-1), coloca "a" antes que "b"
Un 0, los deja como están.

Haciendo uso de esa capacidad podemos hacer ordenaciones complejas o personalizadas.